extensions/resource-deployment/src/helpers/optionSources.ts

```diff
@@ -11,74 +11,112 @@ import { apiService } from '../services/apiService';
 import { throwUnless } from '../utils';
 import { CacheManager } from './cacheManager';
 
-export enum OptionsSourceType {
-	ArcControllersOptionsSource = 'ArcControllersOptionsSource'
+export const enum OptionsSourceType {
+	ArcControllersOptionsSource = 'ArcControllersOptionsSource',
+	ArcControllerConfigProfilesOptionsSource = 'ArcControllerConfigProfilesOptionsSource'
 }
 
-export abstract class OptionsSource implements IOptionsSource {
+export namespace OptionsSources {
```

Comment:
Why are you creating a new namespace? That seems unnecessary

Reply:
Is it just so you can do `return new OptionsSources[sourceType](variableNames, sourceType);`?

I don't know if this is a common pattern - I certainly haven't seen it before. But I really don't like doing needlessly fancy stuff like this when you could just simply do a switch statement which is clearer and easier to understand IMO.

So I'm going to ask that you just remove this whole create function and add another line to the switch I had before. There was nothing wrong with that approach and this isn't gaining us anything useful.

Reply:
For a specific example of why I disagree with this being a good change - unless I'm mistaken you lose all type validation for the constructor by indexing into the namespace like that. And that very much is something we should always avoid doing.

Reply:
yes, it so we can do return new OptionsSources[sourceType](variableNames, sourceType);?

And there are times when we ca to any, so what you worry about is not a concern in my mind. The caller gets full type safety, and it is just this one line of code and I reckon this kind of code would be needed when tackling #12682. I do not like the massive switch statement in modelViewUtils. Switch statements add to code complexity (fewer branches in code is good for code maintainability and switch is the mother of branches.

In my view, we can handle this as appropriate when we do #12682. But I would like it if we kept the create method here as I did not invent this factory method and it not uncommon in javascript/typescript world.

Reply:
I don't understand your comment about casting to any. We never do that for constructors - and the times we do it for anything else are fairly rare and usually indicate a code smell.

If we update the constructor to require an additional parameter then this will just silently be broken unless we also remember to update the function call - that's bad and would never happen with the switch statement.

There are ways to get type safety for this kind of thing - see what VS Code does with the instantiation service. I haven't looked into the implementation of that to know exactly what they're doing there and regardless that's way overly complex for this one-off case.

This is not providing anything useful and is just adding extra lines of code and complexity for no real benefit.

The switch statement will go away with the linked work item - we'll just have a map of the registered callbacks instead. So none of this is long lasting anyways.

Reply:
reverting ...